src/lesson_4/FractionalWrapper.sol

```diff
@@ -152,7 +152,15 @@ contract FractionalWrapper is ERC20 {
     ///@param owner the owning address
     ///@return shares amount of vault asset
     function withdraw(uint256 assets, address receiver, address owner) public returns (uint256 shares) {
-        shares = _convertToShares(assets);
+        shares = previewWithdraw(assets);
+
+        if (msg.sender != owner) {
+            uint256 allowed = _convertToShares(_allowance[owner][msg.sender]);
+            require(allowed > 0, "Insufficient allowance");
```

Comment:
This is a bit of a confusing check, because you could have insufficient allowance despite it not being zero, and you'll get a panic instead of this require.

I would either remove the require, or change it to `allowed > shares`.

Reply:
Odd. I removed these checks in my last commit but it's showing in the review. But not in the changed files.